### PR TITLE
8294997: Improve ECC math operations

### PR DESCRIPTION
**Pull request by djelinski**
This patch rewrites some BigInteger and curve point operations used in EC calculations:
- coefficient * 2^power is equivalent to coefficient << power
- number mod 2^n is equivalent to number & (2^n-1)
- pair of IntegerModuloP operations:
t2 = t1+t1
t1 = t1+t2
is equivalent to t1=t1*3, which is now implemented more efficiently.

Benchmarked the code using not-yet-merged benchmark from #10544. Results on x64 before:
```
Benchmark        (messageLength)   Mode  Cnt     Score   Error  Units
Signatures.sign               64  thrpt   15  1578.907 ± 1.522  ops/s
```
After:
```
Benchmark        (messageLength)   Mode  Cnt     Score   Error  Units
Signatures.sign               64  thrpt   15  1679.495 ± 3.883  ops/s
```
Greatest part of the improvement is related to ECOperations changes; BigInteger modifications provide only marginal gains (1584 ops/s without ECOperations changes).

Tier1-3 tests continue to pass.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294997](https://bugs.openjdk.org/browse/JDK-8294997): Improve ECC math operations

### Reviewers
 * [Xue-Lei Andrew Fan](https://openjdk.org/census#xuelei) (@XueleiFan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10614/head:pull/10614` \
`$ git checkout pull/10614`

Update a local copy of the PR: \
`$ git checkout pull/10614` \
`$ git pull https://git.openjdk.org/jdk pull/10614/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10614`

View PR using the GUI difftool: \
`$ git pr show -t 10614`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10614.diff">https://git.openjdk.org/jdk/pull/10614.diff</a>

</details>

### CODE REVIEW

**Comment by ferakocz** [comment]
In src/java.base/share/classes/sun/security/provider/DSAParameterGenerator.java there a still some mod(Biginteger.TWO.pow(...)) type subexpressions, e.g. in lines 214, 220, 231, 240 (not that it matters much).

File: src/java.base/share/classes/sun/security/provider/DSAParameterGenerator.java [hunk 1]
```diff
@@ -235,15 +235,15 @@ private static BigInteger[] generatePandQ(SecureRandom random, int valueL,
                 /* Step 11.2 */
                 BigInteger W = V[0];
                 for (int i = 1; i < n; i++) {
-                    W = W.add(V[i].multiply(BigInteger.TWO.pow(i * outLen)));
+                    W = W.add(V[i].shiftLeft(i * outLen));
                 }
                 W = W.add((V[n].mod(BigInteger.TWO.pow(b)))
```
Comment (XueleiFan) [inline src/java.base/share/classes/sun/security/provider/DSAParameterGenerator.java:240]: Did you want to update the "BigInteger.TWO.pow(b)" as well?

File: src/java.base/share/classes/sun/security/util/math/intpoly/IntegerPolynomial.java [hunk 2]
```diff
@@ -329,9 +329,10 @@ private void setLimbsValue(BigInteger v, long[] limbs) {
     }
 
     protected void setLimbsValuePositive(BigInteger v, long[] limbs) {
-        BigInteger mod = BigInteger.valueOf(1 << bitsPerLimb);
+        assert bitsPerLimb < 32;
```
Comment (XueleiFan) [inline src/java.base/share/classes/sun/security/util/math/intpoly/IntegerPolynomial.java:332]: I may have this assert in the constructors as it is a final field.

File: src/java.base/share/classes/sun/security/provider/DSAParameterGenerator.java (hunk 1 above)
Comment (djelinski) [reply src/java.base/share/classes/sun/security/provider/DSAParameterGenerator.java:240]: Actually I think I'll revert the DSA changes; I'm not prepared to measure their effect at this moment.

File: src/java.base/share/classes/sun/security/util/math/intpoly/IntegerPolynomial.java (hunk 2 above)
Comment (djelinski) [reply src/java.base/share/classes/sun/security/util/math/intpoly/IntegerPolynomial.java:332]: I added it here as a comment explaining why using intValue is okay here; it wouldn't serve its purpose if I moved it elsewhere.
Asserts are eliminated by the compiler by default anyway. If you think this assert shouldn't be here, I can replace it with a normal comment.

Comment (XueleiFan) [reply src/java.base/share/classes/sun/security/util/math/intpoly/IntegerPolynomial.java:332]: The assert feature can be turned on.  I may prefer to use comment if that's the intention.

**Review by XueleiFan — APPROVED** [review]

**Pull request closed without merging**